### PR TITLE
MNT/TST: Update to Python 3 and Add Tests

### PR DESCRIPTION
**Pull request by ZLLentz**
This is now ready for reviews. I'm asking more people than I normally would because this is used in a lot of places, but if you have other things to work on you're under no obligation to review.

## Things to focus on as a reviewer
- Are there any new red flags in expand.py?
- Does the testing seem sufficient?
- Is there anything else I should document better while I'm here?

## Background

`ioc-template-macros` is a repo that provides the following tools for building templated IOCs:
- `RULES_EXPAND`: a file to include in a templated IOC `Makefile` such that when we `make` that IOC it uses this repo to expand the templates. A typical templated IOC `Makefile` is something like:
```
# SLAC PCDS Makefile for building templated IOC instances
IOC_CFG  += $(wildcard *.cfg)
include /reg/g/pcds/controls/macro/RULES_EXPAND
```
- `expand`: a shell script that sets up a Python environment to run `expand.py`
- `expand.py`: a python script that reads from config files and uses the data within them to expand IOC templates.

## Main Goal of PR

The main goal of this PR is to update `expand.py` from python 2 to python 3 and switch out the `expand` script for one that references our python 3 `conda` environments instead of the old python 2 `pspkg` environments.

This will:
- Allow us to build templated IOCs on the rocky9 hosts without rebuilding Python 2, etc. on the new arch
- Allow us to move on from Python 2, which has been fully discontinued since the start of 2020.

Of note: I made no effort to refactor the existing code. I am aware that the code is messy, confusing, and difficult to maintain. I am aware that it doesn't follow any of our standard deployment strategies. This is a task for another day.

## How it was done

To accomplish the main goal, `expand.py` was automatically updated using:
- 2to3
- ruff --fix
- ruff format
- manual edits until `ruff` was happy

And then, afterwards, there was one remaining bug where the `TRANSLATE` macro was broken due to an API change outside of the normal 2to3 scope. I fixed this bug.

I also used some of my learning from this process to convert `NOTES` to a `README.md` and add some useful information.

## Unit Tests

The rest of the effort here is related to testing, since I need to ensure that this sweeping update doesn't break existing IOCs. Here's an overview of the test infrastructure:
- `expand.py` was modified to create `main` and make it importable for testing.
- `tests/generate_artifacts.py` fills the `tests/examples`, `tests/common`, and `tests/expected` directories with real outputs from `expand.py` from past builds.
- These folders are not committed because there are thousands of files and it makes the PR impossible to review. They can be added in a follow-up PR.
- `tests/examples` contains `.cfg` files that `expand.py` consumes to get data to use to fill templates.
- `tests/common` contains the `iocBoot/template` directories from the parent (common) iocs referenced in the `tests/examples`
- `tests/expected` contain the fully expanded files from the most recent past builds
- `tests/test_expand.py` runs more than 4000 individual tests based on the contents of these folders, where each test calls `main()` from `expand.py`
- This file also runs a few checks that the expand preprocessing still works
- `tests/test_keywords.py` was intended to hold unit tests for specific special keywords from `expand.py` but I ended up only writing 2 of these tests
- `tests/test_rules_expand.py` checks that the `RULES_EXPAND` Makefile from the repo root is still properly usable when you `make` your template IOC. It accomplishes this by using a dummy IOC whose internal build does not rely on our local EPICS build system at all. It uses `tests/ioc-tst-unittest` for this.

Offline, I get the following test results:
```
4260 passed in 368.45s (0:06:08)
```

Note that this test suite takes a long time to run because `expand.py` is really _really_ slow at expanding `ioc-common-pvNotepad`'s `st.cmd` file.

### CODE REVIEW

**Comment by ZLLentz** [comment]
For the purposes of this PR I am doing the following:

- Putting the contents of the archive folders into .gitignore
- Purging the contents of the archive folders from the commit history
- Force pushing back to this branch

So that it will be easier to review the result.

There will be a follow-up PR which simply:

- Removes the .gitingore entry
- Re-fills the archive folders

**Comment by ZLLentz** [comment]
Hmm I broke it, so instead of rewriting the history the "clever" way I'm doing it the "simple" way (not preserving my long chains of commits)

**Comment by ZLLentz** [comment]
What went wrong is that I ended up rewriting the commits from prior to this PR in the history, which makes it look really strange here and is just overall annoying.

**Comment by ZLLentz** [comment]
Including "common" added about 1000 test cases (3261 -> 4286) and some of these are still failing. At least the first remaining one looks like a classic "they edited the built IOC", some of them look like "used an alternate but equivalent filepath", etc. but I'll continue investigating tomorrow. The approach is to check them one by one and either fix the associated expand.py bug or ban the offending IOC.

**Comment by ZLLentz** [comment]
This is now ready for reviews. I'm asking more people than I normally would because this is used in a lot of places, but if you have other things to work on you're under no obligation to review.

File: tests/ioc-tst-unittest/Makefile
Comment (ddamiani) [inline tests/ioc-tst-unittest/Makefile:1]: If I'm remembering correctly the first character of lines that are recipes (a.k.a shell commands that get run) need to be denoted by a single tab character. It's some quirk of how the makefile parser works...

Also tabs are not useable on lines where you are just defining variables. I believe there is a way to override the magic recipe character to be something other than tab if you want.

**Comment by ddamiani** [comment]
Looks good to me the changes to the code itself look fine. Just straight forward changes for going from 2 to 3 and basic linter/code format fixes.

Testing strategy looks good.

**Review by ddamiani — DISMISSED** [review]
Looks good to me the changes to the code itself look fine. Just straight forward changes for going from 2 to 3 and basic linter/code format fixes.

Testing strategy looks good.

File: expand.py
Comment (tangkong) [inline expand.py:1]: In a world where we have infinite time, would we benefit from transitioning to some sort of grammar parsing (lark)?

**Comment by janeliu-slac** [comment]
Looks good, not seeing any issues so far. Curious what is the abstract syntax tree used for in expand.py?

File: tests/generate_artifacts.py [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+            logger.info(
+                f"In {ioc_path} found {version_path.name} "
+                "which has fewer than 2 elements."
+            )
+            continue
+        if version > latest:
+            latest = version
+            latest_path = version_path
+    if latest_path is None:
+        raise RuntimeError(f"No version directories in {ioc_path}")
```
Comment (tangkong) [inline tests/generate_artifacts.py:204]: Nit: This feels more like a [`FileNotFoundError`](https://docs.python.org/3/library/exceptions.html)

File: tests/ioc-tst-unittest/Makefile
Comment (ZLLentz) [reply tests/ioc-tst-unittest/Makefile:1]: This is all true and gave me a headache when getting this dummy Makefile to work. These needed be runnable on CI such that the fake IOC would complete a build.

**Comment by ZLLentz** [comment]
> Curious what is the abstract syntax tree used for in expand.py

This made me curious as well when I was working through the PR.
It turns out that there are special `CALC` and `IFCALC` macros that evaluate expressions as python code.

**Review by tangkong — COMMENTED** [review]
The brain is petering out, I'll revisit again tomorrow.  Broadly I also think the testing strategy makes sense, and am looking forward to however you decide to upload all 5000+ test artifacts 😆

File: expand.py
Comment (ZLLentz) [reply expand.py:1]: I think this makes sense in principle, but I'd rather spend my infinite time making it possible to template our IOCs using jinja2!

**Comment by ZLLentz** [comment]
> looking forward to however you decide to upload all 5000+ test artifacts

I decided to count them up, it turns out that there are 10853 test artifacts with a combined file size of about 67MB.

**Comment by tangkong** [comment]
That's not quite git LFS status yet is it?

**Review by JJL772 — DISMISSED** [review]
Changes look good and work well with ioc-common-bk-1697. Always nice to see less Python 2!

**Comment by ZLLentz** [comment]
> That's not quite git LFS status yet is it?

At that scale, I think the appropriate way to include them will be in a separate PR, including the files as-is.

**Review by tangkong — APPROVED** [review]
I'm excited to see the followup PR

**Pull request merged**